### PR TITLE
implemented the file creation in the destination folder

### PR DESCRIPTION
**Pull request by ibilalkayy**
Hey there,

In this PR, I implemented the functionality of file creation in the destination folder by taking the text from the original files and copying it to the destination files

The part of removing the original files and running the updated code still remains because I didn't change the main file for now.

You can check and let me know about the improvements that need to be made.

Thank you!

### CODE REVIEW

**Review by nitnelave — CHANGES_REQUESTED** [review]
Before drilling down into the details, I have a few high-level comments:
 - This looks like a high-impact PR: it changes or defines the command-line behavior of the whole program. In addition, it introduces a lot of logic that would need to be validated. Usually, we try to separate the two: either add internal details, well-tested logic OR change the visible behavior of the program (e.g. by using previously introduced logic). A good rule of thumb is to "first make the change easy, then make the change"
 - this PR doesn't have any test. How do I know the code does what you expect it to? I would expect any non-trivial PR to include unit tests covering not only the happy path, but also all the edge cases. (Note that I saw several bugs in the code)
 - finally, in terms of communication, this doesn't implement what we talked about. Your change seems to be mostly about finding which file(s) to change and copying them to the right place. I was suggesting to implement replacing some parts of the content of a file, given the source and destination (so you don't have to worry about that part).

File: cli/src/main.rs
Comment (nitnelave) [inline cli/src/main.rs:1]: Can you remove the changes from this file?

File: cli/src/dependencies.rs
Comment (nitnelave) [inline cli/src/dependencies.rs:1]: And this file as well

File: cli/src/transform.rs [hunk 1]
```diff
@@ -0,0 +1,19 @@
+use std::{
+    fs::{File, read_to_string},
+    path::Path,
+};
+
+pub fn transform(input_path: &str, output_path: &str, use_statements: &str) {
```
Comment (nitnelave) [inline cli/src/transform.rs]: This function doesn't do what you think it does. Not only that, but you also have to convince me that it does what I want it to do! The best way to do this is to add tests.

Also, we can improve the signature: the file paths can be `std::path::Path`, and the use statements should be a `UseStatements` from the `dependencies` module of the `extricrate` crate (in the crates folder)

By the way, could you put that function inside the `extricrate` crate? (Not the main cargo-extricrate binary)

**Review by nitnelave — CHANGES_REQUESTED** [review]

**Comment by ibilalkayy** [comment]
Hey there, I made modifications by giving the module, crate_name as paths and used the UseStatement by taking the data from the lib.rs file except span because I didn't understand what it does. I didn't write tests for it, but you can look at the changes and see if I am on the right track. Thank you!

File: crates/extricrate/src/lib.rs [hunk 2]
```diff
@@ -76,12 +76,12 @@ pub mod dependencies {
     #[derive(Debug)]
     pub struct UseStatement {
         /// Where the use statement appears.
-        source_module: ModuleName,
+        pub source_module: ModuleName,
         /// List of referenced modules.
         /// Several targets, to represent `use crate::{log, foo::{bar, baz}};`
-        target_modules: HashSet<ModuleName>,
+        pub target_modules: Vec<ModuleName>,
```
Comment (nitnelave) [inline crates/extricrate/src/lib.rs]: Why did you change this?

**Review by nitnelave — CHANGES_REQUESTED** [review]
The signature of the `transform` function is correct, but I don't want to put code in the main crate yet. I would like all the changes in this PR to be limited to the "extricrate" crate. All the changes should be in the crates/extricrate directory.

File: crates/extricrate/src/lib.rs (hunk 2 above)
Comment (ibilalkayy) [reply crates/extricrate/src/lib.rs]: They are the private fields that were not able to be accessed in the main.rs

Comment (nitnelave) [reply crates/extricrate/src/lib.rs]: Yes, I see the addition of `pub` everywhere (though I have my own thoughts about that).

But why a Vec instead of a HashSet?

Comment (ibilalkayy) [reply crates/extricrate/src/lib.rs]: Here is the answer: [link](https://github.com/nitnelave/extricrate/commit/341526752da14e5d7e3f43b733406b026133636e#r159521011)

Comment (nitnelave) [reply crates/extricrate/src/lib.rs]: Yes, that's why we switched from a Vec to a HashSet, but you're undoing this change. I assume it was accidental?

Comment (ibilalkayy) [reply crates/extricrate/src/lib.rs]: Yeah it was accidental.

**Review by nitnelave — CHANGES_REQUESTED** [review]
Can you have a look at my open comments? Once you've addressed them I'll take another look. Not much seems to have changed since last time I reviewed.

**Comment by ibilalkayy** [comment]
Hey there,

I moved the transform function in the lib.rs file, and worked on it in such a way that it took input and output paths, took the commands from the user, and replaces the use statements with todo. This was all done without using the use_stateaments as a function parameter but it worked.

Then I tried to use statements and confused about the source, target, module_name and statement_type and span(yea it locats the use statements) but I didn't understand the implementation of it although I used the same use_item in the above it to make it work and it worked.

**Comment by nitnelave** [comment]
Can you try addressing all the open comments, or at least replying to them, before requesting a new review?

File: cli/src/main.rs
Comment (ibilalkayy) [reply cli/src/main.rs:1]: I removed the changes from this file.

File: cli/src/transform.rs (hunk 1 above)
Comment (ibilalkayy) [reply cli/src/transform.rs]: I removed this file and used functionality in the lib.rs but I am not sure about the use statements as I have implemented the input and output in the function parameters but confused about the UseStatements and it's usage.

Although the text was still replaced with todo without using the use statement

File: cli/src/dependencies.rs
Comment (ibilalkayy) [reply cli/src/dependencies.rs:1]: I removed this file and used functionality in the lib.rs

File: crates/extricrate/src/lib.rs (hunk 2 above)
Comment (ibilalkayy) [reply crates/extricrate/src/lib.rs]: I pulled the changes and now the file is having the HashSet instead of the Vec.

Comment (ibilalkayy) [reply crates/extricrate/src/lib.rs]: Plus all the public keywords are removed because all the functionalities are used in the lib.rs file.

**Comment by ibilalkayy** [comment]
Sorry I missed them but now I have replied to all of them and you can take a look.

File: Cargo.toml [hunk 3]
```diff
@@ -1,8 +1,5 @@
 [workspace]
-members = [
```
Comment (nitnelave) [inline Cargo.toml:2]: I don't think this is needed either

File: cli/Cargo.toml [hunk 4]
```diff
@@ -17,6 +17,9 @@ tracing-attributes.workspace = true
 tracing-forest.workspace = true
 tracing-log.workspace = true
 tracing-subscriber.workspace = true
+extricrate = { path = "../crates/extricrate" }
```
Comment (nitnelave) [inline cli/Cargo.toml:20]: None of these should be necessary

File: cli/src/main.rs [hunk 5]
```diff
@@ -1,5 +1,7 @@
 #![allow(dead_code, unused_variables)]
 
+use std::path::Path;
```
Comment (nitnelave) [inline cli/src/main.rs]: There's still this line which doesn't belong

File: crates/extricrate/src/lib.rs [hunk 6]
```diff
@@ -30,8 +30,8 @@ pub mod dependencies {
     /// A single, separate use statement.
     #[derive(Debug, PartialEq, Eq)]
     pub struct NormalizedUseStatement {
-        module_name: ModuleName,
-        statement_type: UseStatementType,
+        pub module_name: ModuleName,
```
Comment (nitnelave) [inline crates/extricrate/src/lib.rs:33]: Do you need this? If you do, it should at most be `pub(crate)`

File: crates/extricrate/src/lib.rs [hunk 7]
```diff
@@ -86,6 +86,57 @@ pub mod dependencies {
...
     pub type UseStatements = Vec<UseStatement>;
 
+    pub fn transform(input_path: &Path, output_path: &Path, use_statements: UseStatements) {
+        // Check whether the output path exists or not
+        if !output_path.exists() {
+            FsFile::create(output_path).expect("Err: failed to create a file");
+        }
+
+        // Read the input path content
+        let content = read_to_string(input_path).expect("Err: failed to read the file content");
```
Comment (nitnelave) [inline crates/extricrate/src/lib.rs]: It might be easier to manipulate bytes than UTF-8 strings (see my later comments)

File: crates/extricrate/src/lib.rs [hunk 8]
```diff
@@ -86,6 +86,57 @@ pub mod dependencies {
...
+        let syntax: SynFile = syn::parse_file(&content).unwrap();
```
Comment (nitnelave) [inline crates/extricrate/src/lib.rs]: You shouldn't need to parse the file. The parsing was done in advance, and you should have all the information you need in the UseStatements.

The UseStatements gives you a list of statements to replace with "todo". The most interesting part right now is the span: it gives you the beginning and end (in bytes) of the statement in the input file.

The algorithm I had in mind was along these lines:
 - copy the content of the file up to the beginning of the first use statement, write it at the destination.
 - now you are at the point where the use statement was, so you can write "todo"
 - after that, copy the input from after the statement until the beginning of the next statement
 - repeat until you cover all the statements 

(Obviously this is not even pseudo code, the code structure will be different)

And don't forget to add tests to make sure it works!

File: crates/extricrate/src/lib.rs [hunk 9]
```diff
@@ -86,6 +86,57 @@ pub mod dependencies {
 
     pub type UseStatements = Vec<UseStatement>;
 
+    pub fn transform(input_path: &Path, output_path: &Path, use_statements: UseStatements) {
```
Comment (nitnelave) [inline crates/extricrate/src/lib.rs]: Can you try to find a better name for the function?

And put it in a different module (maybe the module can be called "transform")

**Review by nitnelave — CHANGES_REQUESTED** [review]

File: Cargo.toml (hunk 3 above)
Comment (ibilalkayy) [reply Cargo.toml:2]: Then what should be added.

File: cli/Cargo.toml (hunk 4 above)
Comment (ibilalkayy) [reply cli/Cargo.toml:20]: I think you're right that I am not using it in the binary, so I am going to remove it.

File: cli/src/main.rs (hunk 5 above)
Comment (ibilalkayy) [reply cli/src/main.rs]: I forgot to remove but now it is removed.

File: crates/extricrate/src/lib.rs (hunk 6 above)
Comment (ibilalkayy) [reply crates/extricrate/src/lib.rs:33]: the transform function is in the lib.rs fie, so I don't think I need pub for now. It maybe due to the mistake

File: Cargo.toml (hunk 3 above)
Comment (nitnelave) [reply Cargo.toml:2]: I mean you don't need to change this file. This is just a formatting change, it doesn't belong in this PR. And I don't think it is better than before.

**Comment by ibilalkayy** [comment]
@nitnelave  I have submitted the code where I successfully replaced the `use std::path::Path` with `todo()!` and I created a new module for it in the library crate and it's now working.

I didn't write the tests yet but I want the feedback if the improvement is required.

File: crates/extricrate/src/lib.rs (hunk 8 above)
Comment (nitnelave) [reply crates/extricrate/src/lib.rs]: @ibilalkayy this is the key comment, that should give you a good idea of how to implement the PR.

Please start by reducing the size of the PR. Remove all non-relevant changes. Your PR should _only_ be creating the "transform" module and the new function inside (plus tests).

I want to repeat: this function should _not_ do any parsing. The parsing was already done, you have all the information you need in `UseStatements`.

Comment (ibilalkayy) [reply crates/extricrate/src/lib.rs]: @nitnelave  I just made it simplified again, but I am confused about the tests that are going to be written, because in this [newly pushed code](https://github.com/nitnelave/extricrate/pull/4/commits/fbdc5f728af81cc12eedab34c489348febb74e4a), I followed the structure of the struct and that's it. Now new values are going to be added and without replacing it with todo(), how can I write tests?

I don't know if you can tell me in easy words. Sometimes it's hard for me to grasp. Maybe it's because of not understanding the struct fields properly.

My understanding is following. Please correct me if I am wrong.

- ModuleName = It shows the path of a module. crate::utils::math::add, crate::utils::math is the path and add is the function name or type
- StatementType = Simple, Alias or Wildcard
- UseStatement = Use line with extra info
	1. source_module → where in the project this use statement lives.
	Example: if the file is src/foo.rs, source_module might be "crate::foo".
	2. target_modules → all the modules it imports.
	Example:
		use crate::{log, foo::{bar, baz}};

	Target modules here would be:
	"crate::log"
	"crate::foo::bar"
	"crate::foo::baz"
- UseStatementDetail
	This is like a zoomed-in view of the use line.
	It usually contains:
	items → a list of individual use parts (each with its module name and type: simple, alias, or wildcard).
	span → where in the source file the use statement is located (start and end position).

Comment (nitnelave) [reply crates/extricrate/src/lib.rs]: The idea of the existing code is to parse a given file and create the list of imports/use (call it whatever you want) in the simplest way possible. We ignore anything that's not a "use" statement. So for instance, if you have a given file `foo.rs` that starts with:

```rust
use std::collections::HashMap;
use super::my_module::{self, NewType, submodule::OtherType};
use chrono::*;

// ...
```

We would generate a list of use statements that looks roughly like:
 - module_name: "std::collections", statement_type: Simple("HashMap")
 - module_name: "crate::my_module", statement_type: Simple("self")
 - module_name: "crate::my_module", statement_type: Simple("NewType")
 - module_name: "crate::my_module::submodule", statement_type: Simple("OtherType")
 - module_name: "chrono", statement_type: WildCard

That's the `NormalizedUseStatement` (from `flatten_use_tree`).

What you're going to work with is the `UseStatement` which contains both this and the syntax details of how the code was written. For the same example, we'll have something like a list of UseStatement:
 - source_module: `crate::foo` (because it's in `foo.rs`), target_modules: ["std::collections"], statement: { items: [{module_name: "std::collections", statement_type: Simple("HashMap")}], span: first line}
 - source_module: `crate::foo`, target_modules: ["crate::my_module", "crate::my_module::submodule"], statement: {items: [ the three normalized statements described above, all starting with "crate::my_module"], span: the second line}
 - a third entry for the chrono import.

To play around with the code, you can check the test `aliases_resolved`. You can copy the part up to `visitor.visit_file(&file)`, change the source text, and then look at what's inside `visitor.use_statements`.

As for the types, it seems you understood what they were used for. I'll add a few notes:

`ModuleName` is supposed to always be a module. (note @ksnll that Bilal pointed out a bug: we use the case of the last element of a "use" statement to determine whether it's a module or not, but that won't work if we're importing a function. We might have to try to resolve the module and fall back to the parent if it doesn't exist). Your example is correct "crates::util::math" would be the module name, though currently due to a bug we would consider that "add" is also a module.

In `UseStatement`, the target modules are only the actual modules, not elements. So if we import "crate::math::{Number, ComplexNumber}", we would only have "crate::math" in the target modules.

The `UseStatementDetail` contains the normalized imports, one per item (so the group imports are separated into single imports).

So it seems you have a good grasp of the types of the project! For the function you need to implement, however, you won't be using most of that: basically you only need the `span` for now. Later on, as we'll add more complexity, we'll make use of more fields.

**Comment by ibilalkayy** [comment]
@nitnelave I modified the function this time and wrote one of the test also that you suggested and it started working by showing the use statement type with span. You can check and tell me the improvements or further steps.

**Comment by nitnelave** [comment]
The test is a good template, you got the list of use statements. Now you need to pass that to a function that will replace them with todo, and check that the resulting text is as expected.

Of course, you'll need to write that function.

**Comment by ibilalkayy** [comment]
@nitnelave I am just trying to understand, how that will look like because my approach was beginner friendly. I was directly trying to write the logic without writing tests but this approach seems strange to me that everything is done in testing at first.

I changed the function name to `find_use_statement` and the second function is going to be `replace_use_statement` but I don't understand that how it will take the `src` input (I know that it's dummy) because in the `find_use_statement`, I followed a structure to reach to a span and tested it but here I am not finding any kind of structure to follow and replace it with todo.

If I try to perform the logic here, then I have to repeat the following code in the `replace_use_statement` function also which is against `DRY`.

```
let module_name = ModuleName("crate::foo".to_string());
let statement_type = UseStatementType::Simple("std::path".to_string());
let normalized_use_statement = NormalizedUseStatement {
    module_name: module_name.clone(),
    statement_type: statement_type,
};

let data = UseStatement {
    source_module: module_name.clone(),
    target_modules: HashSet::from([module_name.clone()]),
    statement: UseStatementDetail {
        items: vec![normalized_use_statement],
        span: Span::call_site(),
    },
};
```

**Comment by nitnelave** [comment]
@ibilalkayy maybe try to do an example by hand: what does the input look like, what will the output look like? 

Once you have a basic example like that, you can write the code to make it match, and then think about other tests

**Comment by ibilalkayy** [comment]
@nitnelave, this time I added the `replace_use_statement` function that prints the whole path of simple, alias, or wildcard statement type but I also used some of the finder function code in the replace function.

**Comment by nitnelave** [comment]
@ibilalkayy I'm not sure we're on the same page. Can you explain in your own words what you're trying to write? Can you give an example input and output for the function?

**Comment by ibilalkayy** [comment]
@nitnelave I pushed the code where the `module_name` and the `statement_type` are concatenated and then printed. After this printing, I am thinking of replacing it with todo and then writing the tests.

Here is the code below that I pushed recently to replace it, but for now, it is just printing the concatenated path. Some of the code is repeated in this function that was also written in the previous function of `find_use_statement`

Let me know if this is the correct approach.

```
pub fn replace_use_statement() {
        // repeated code
        let module_name = ModuleName("crate::foo".to_string());
        let statement_type = UseStatementType::Alias("std::path".to_string(), "mypath".to_string());
        let normalized_use_statement = NormalizedUseStatement {
            module_name: module_name.clone(),
            statement_type: statement_type,
        };

        let data = UseStatement {
            source_module: module_name.clone(),
            target_modules: HashSet::from([module_name.clone()]),
            statement: UseStatementDetail {
                items: vec![normalized_use_statement],
                span: Span::call_site(),
            },
        };
        
        // new code
        let statement_data = match &data.statement.items[0].statement_type {
            UseStatementType::Simple(s) => s.clone(),
            UseStatementType::Alias(s, alias) => format!("{} as {}", s.as_str(), alias.as_str()),
            UseStatementType::WildCard => "*".to_string(),
        };

        println!(
            "Use statement: {}::{}",
            data.statement.items[0].module_name.0, statement_data
        );
    }
```

**Comment by nitnelave** [comment]
Rather than directly iterating on the code, I'd rather start by making sure we agree on what this PR is about. Can you summarize what you think the PR should do?

**Comment by ibilalkayy** [comment]
As far as I have understood, you want me to find the use statement that will take the module name and statement type and replace it with todo.

For now, I have found the use statement, wrote tests but now the next step is replace it with todo(that I have just printed for now) and write test for it.

Obviously, this is some dummy data that belongs to variables only in the function and that data is not taken from the CLI commands that the user enters, but the real goal is to replace the actual data from the commands, find it and then replace it with todo.

Tell me, if this thing matches your approach about this PR and do we agree

**Comment by nitnelave** [comment]
The goal of this function is to take _all_ the use statements and replace them with "todo".

The next step (next PR) is to replace only the statements that import a module or its given submodules (it has interesting problems, for instance with statement groups).

The step after that is to generate the correct replacement imports (targeting the new crate we extract instead).

And that last step can be directly integrated in the overall extraction process.

**Comment by ibilalkayy** [comment]
What I am doing is the first step which is to replace the use statements with todo. As the function is submitted, I wanted to get the feedback whether it is right or needs improvement to replace it with todo because currently it is taking the module name and the statement and concatenate them and then print them.

**Comment by nitnelave** [comment]
@ibilalkayy what does the function operate on? What are the inputs?

Currently, your PR adds 2 functions that take no inputs.

**Comment by ibilalkayy** [comment]
@nitnelave I understood your point that I have to take the input as a module name and a crate name from the terminal and apply the functionality of replacing the use statement with real data.

Here is the struct that is present in the `main.rs` file, which works as inputs:

```
#[derive(Debug, Parser, Clone)]
pub struct ExtractOpts {
    /// Module to extract from a crate.
    #[clap(long, env = "EXTRICRATE_MODULE")]
    pub module: String,
    /// Target crate to create.
    #[clap(long, env = "EXTRICRATE_CRATE_NAME")]
    pub crate_name: String,
}
```

Now tell me two things.

1. My previous experience tells me of creating this struct again with the same two fields in a library file and apply a method to perform the logic of `replace_use_statement` method, just like I did it with manual data.

2.  In the above ExtractOpts struct, it is asking two inputs from the user. The first one is the `module(or module_name)` and the second one is the `crate_name(which basically shows the target module, right?)` but from the following code, you can see that the concatenation happens b/w the module and the statement type, not with the target_module or crate_name.

This thing confuses me about whether the below code needs to be imporved or th flag is wrong.

```
 let module_name = ModuleName("crate::foo".to_string());
 let statement_type = UseStatementType::Alias("std::path".to_string(), "mypath".to_string());
 let normalized_use_statement = NormalizedUseStatement {
     module_name: module_name.clone(),
     statement_type: statement_type,
 };

let data = UseStatement {
    source_module: module_name.clone(),
    target_modules: HashSet::from([module_name.clone()]),
    statement: UseStatementDetail {
        items: vec![normalized_use_statement],
        span: Span::call_site(),
    },
};

let statement_data = match &data.statement.items[0].statement_type {
    UseStatementType::Simple(s) => s.clone(),
    UseStatementType::Alias(s, alias) => format!("{} as {}", s.as_str(), alias.as_str()),
    UseStatementType::WildCard => "*".to_string(),
};

println!(
    "Use statement: {}::{}",
     data.statement.items[0].module_name.0, statement_data
);
```

**Note:** I know that this issue is taking so much time but believe me I learned a lot from you. I want to continue and thank you for being patient!

**Comment by nitnelave** [comment]
@ibilalkayy I think there's a misunderstanding: your code right now is not called when running the program. It will never be called directly from main, so the arguments it receives are not the direct user inputs.

The flow of the program goes like this:
- the user asks to extract crate C from module M in the original crate A.
- we parse all of the use statements in A, make a map of the imports, and check whether we can extract the crate (M is self-contained)
- then we move the files from M to a new folder for crate C, and update the local imports in M, as well as the imports of M in A to reflect the new crate (this is where your code comes in)

So your code will receive the already-parsed file with the location of all the modules and instructions on how to update them (which will depend on each use case). For now, we're setting those instructions to be "replace with todo!()". Obviously, once the code is complete (not in this PR) we'll have better functionality.

Right now we're trying to build a small, self-contained piece of code that can take a file and its parsed contents and replace all the "use" statements with todo. So your function should take the input file, the parsed representation and the output file path, and write the replaced version to the output path.

**Comment by ibilalkayy** [comment]
@nitnelave I didn't understand many of the things that you said above, but as far as I have understood, I implemented based on my understanding.

I removed the duplication of logic in the `replace_use_statement` and worked on the functionality of replacing it with todo and then I called both of the functions in the main function by using the `ExtractOpts` fields and calling them in the find and replace use statement functions.

You can check out and let me if I'm in the right direction.

Thank you!
